ChemMaster ECS

0x6273 opened the pull request:
Logic was moved from the component to a system. The component now only holds data. Lots of code was refactored, with a few minor functional changes:
- Fix bug where beaker appearance wouldn't update after discarding reagents.
- Make container and buffer reagents display alphabetically in the UI.
- Add `DataField`s for controlling how many pills/bottles can be produced at the same time.
- Label only gets reset when the buffer reagents change.

mirrorcult commented on Content.Client/Chemistry/UI/ChemMasterBoundUserInterface.cs:
```diff
@@ -36,20 +36,20 @@ protected override void Open()
-            _window.CreatePillButton.OnPressed += _ => PrepareData(UiAction.CreatePills, null, _window.LabelLine, null, _window.PillAmount.Value, null);
-            _window.CreateBottleButton.OnPressed += _ => PrepareData(UiAction.CreateBottles, null, _window.LabelLine, null, null, _window.BottleAmount.Value);
+            // Setup static button actions.
+            _window.EjectButton.OnPressed += _ => SendMessage(new ItemSlotButtonPressedEvent("container_slot"));
```
this needs to use a static string like it did before or someone is going to forget to change this

mirrorcult commented on Content.Server/Chemistry/Components/ChemMasterComponent.cs:
```diff
@@ -1,362 +1,32 @@
-            }
+        [DataField("pillProductionLimit"), ViewVariables(VVAccess.ReadWrite)]
+        public uint PillProductionLimit = default!;
 
-            if (_bufferSolution?.Contents.Count == 0)
-                _label = "";
+        [DataField("bottleProductionLimit"), ViewVariables(VVAccess.ReadWrite)]
+        public uint BottleProductionLimit = default!;
```
these should either have sensible defaults (like what they're set to in yaml right now) or be marked required because having them default to 0 production doesn't seem like correct behavior

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+        private void UpdateUiState(ChemMasterComponent chemMaster, bool updateLabel = false)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
```
don't get this directly. use the `ItemSlotsSystem` methods

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+                _solutionContainerSystem.TryGetSolution(container.Value, fits!.Solution, out containerSolution);
+            }
+
+            var containerName = CompOrNull<MetaDataComponent>(container)?.EntityName;
```
you don't need to `CompOrNull` metadata, every entity has it

```suggestion
            var containerName = Name(container);
```

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+            }
+
+            var containerName = CompOrNull<MetaDataComponent>(container)?.EntityName;
+            var dispenserName = CompOrNull<MetaDataComponent>(chemMaster.Owner)?.EntityName ?? "";
```
same here

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+                }
+
+                var solution = bufferSolution.SplitSolution(individualVolume);
+                var itemSolution = _solutionContainerSystem.EnsureSolution(item, pills ? "food" : "drink");
```
do not hardcode solution names like this

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+                var itemSolution = _solutionContainerSystem.EnsureSolution(item, pills ? "food" : "drink");
+                _solutionContainerSystem.TryAddSolution(item, itemSolution, solution);
+
+                if (pills)
+                    Comp<SpriteComponent>(item).LayerSetState(0, "pill" + (chemMaster.PillType + 1));
```
`TryComp` this

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+
+                if (label != "")
+                {
+                    var labelComponent = item.EnsureComponent<LabelComponent>();
```
```suggestion
                    var labelComponent = EnsureComp<LabelComponent>(item);
```
don't use the entityuid methods. didn't even know we still had them

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+        private void DiscardReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
```
same here ewith itemslotssystem

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+
+        private void DiscardReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
```
do you really need to ensure the solution every time? just get the solution and early out if it doesn't exist at this point

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+
+        private void OnSetPillTypeMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterSetPillTypeMessage message)
+        {
+            // Ensure valid pill type. There are 20 pills selectable, 0-19.
+            if (message.PillType > 19)
```
this should really be static somewhere. also this doesn't clamp if its <0?

0x6273 replied:
It's an `uint` so it can't be <0.

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+        private void OnCreatePillsMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterCreatePillsMessage message)
+        {
+            // Ensure the amount is valid.
+            if (message.Amount == 0 || message.Amount > chemMaster.PillProductionLimit)
+                return;
```
like above this needs to check if its <0 as well because that would cause bugs

0x6273 replied:
Also an `uint`, can't be <0.

mirrorcult replied:
true

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+
+        private void TransferReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
```
itemslotssystem like before. also do not hardcode this string value, make it a static/const somewhere.

mirrorcult commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -1,20 +1,239 @@
+        [Dependency] private readonly AudioSystem _audioSystem = default!;
+        [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+        [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+        [Dependency] private readonly SharedHandsSystem _handsSystem = default!;
```
```suggestion
        [Dependency] private readonly SharedHandsSystem _handsSystem = default!;

```

mirrorcult reviewed (changes requested):
good PR

why was the container ID even changed from `beakerSlot` to `container_slot`? i dont see a reason for doing it and it means you have to touch maps which is bad

0x6273 commented:
Thanks for the review! I learned a lot from it.

I changed the container ID because the other IDs were snake_case, and also because the word "container" is more fitting since it doesn't just accept beakers.
```yaml
- type: ContainerContainer
    containers:
      beakerSlot: !type:ContainerSlot
      machine_board: !type:Container
      machine_parts: !type:Container
```
I didn't know it would require map changes, so I changed it back.

Also, I have no idea why integration tests are failing. They work on my machine.

mirrorcult commented:
test fail looks erroneous
